crates/router/src/connector/cybersource.rs

```diff
@@ -59,25 +74,101 @@ impl Cybersource {
         connectors.cybersource.base_url.as_ref()
     }
+
+    fn build_error_response(
+        &self,
+        res: Bytes,
+    ) -> CustomResult<types::ErrorResponse, errors::ConnectorError> {
+        let response: cybersource::ErrorResponse = res
+            .parse_struct("Cybersource ErrorResponse")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+        Ok(types::ErrorResponse {
+            code: consts::NO_ERROR_CODE.to_string(),
+            message: response
+                .message
+                .unwrap_or_else(|| consts::NO_ERROR_MESSAGE.to_string()),
+            reason: None,
+        })
+    }
+}
+
+impl<Flow, Request, Response> ConnectorCommonExt<Flow, Request, Response> for Cybersource
+where
+    Self: ConnectorIntegration<Flow, Request, Response>,
+{
+    fn build_headers(
+        &self,
+        req: &types::RouterData<Flow, Request, Response>,
+        connectors: &settings::Connectors,
+    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+        let date = OffsetDateTime::now_utc();
+        let cybersource_req = self.get_request_body(req)?;
+        let auth: cybersource::CybersourceAuthType =
+            cybersource::CybersourceAuthType::try_from(&req.connector_auth_type)?;
+        let merchant_account = auth.merchant_account.clone();
+        let base_url = connectors.cybersource.base_url.as_str();
+        let cybersource_host = Url::parse(base_url)
+            .into_report()
+            .change_context(errors::ConnectorError::RequestEncodingFailed)?;
+        let host = cybersource_host
+            .host_str()
+            .ok_or(errors::ConnectorError::RequestEncodingFailed)?;
+        let path: String = self
+            .get_url(req, connectors)?
+            .chars()
+            .skip(base_url.len() - 1)
+            .collect();
+        let sha256 =
+            self.generate_digest(cybersource_req.map_or("{}".to_string(), |s| s).as_bytes());
+        let http_method = self.get_http_method();
+        let signature = self.generate_signature(
+            auth,
+            host.to_string(),
+            path.as_str(),
+            &sha256,
+            date,
+            http_method,
+        )?;
+
+        let mut headers = vec![
+            (
+                headers::CONTENT_TYPE.to_string(),
+                self.get_content_type().to_string(),
+            ),
+            (
+                headers::ACCEPT.to_string(),
+                "application/hal+json;charset=utf-8".to_string(),
+            ),
+            ("Digest".to_string(), format!("SHA-256={}", sha256)),
+            ("v-c-merchant-id".to_string(), merchant_account),
+            ("Date".to_string(), date.to_string()),
+            ("Host".to_string(), host.to_string()),
+            ("Signature".to_string(), signature),
+        ];
+        if matches!(http_method, services::Method::Get) {
+            headers.remove(2);
+        }
```

Comment:
This would result in the wrong set of headers if someone adds a header at or before index 2. 

Instead, append a `Digest` header if `http_method` is `Post` (or `Put`, if that's also required). 

```rust
if matches!(http_method, services::Method::Post) {
    headers.push(("Digest".to_string(), format!("SHA-256={}", sha256)));
}
```